### PR TITLE
IGNITE-15162 Thin 3.0: Add buffer pooling

### PR DESCRIPTION
**Pull request by ptupitsyn**
* Add server-side and client-side buffer pooling using Netty ByteBuf pooled allocator.
  * https://netty.io/wiki/using-as-a-generic-library.html
  * https://netty.io/wiki/reference-counted-objects.html
* Prepend message length in `ClientMessagePacker` to avoid extra buffer allocation and get rid of `ClientMessageEncoder`.
* Move request handling to separate classes (one class per message type).
* Stop `ConfigurationRegistry` after tests.

### CODE REVIEW

File: modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java [hunk 1]
```diff
@@ -17,48 +17,43 @@
...
     private boolean magicDecoded;
 
     /** Magic decoding failed flag. */
     private boolean magicFailed;
 
+    /**
+     * Constructor.
+     */
+    public ClientMessageDecoder() {
+        super(Integer.MAX_VALUE - HEADER_SIZE, 0, HEADER_SIZE, 0, HEADER_SIZE, true);
```
Comment (ivandasch) [inline modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java:48]: May be it is good idea to limit max message size? 
I have encountered with this issue few days ago. 
There are security scanners, that send random large stuff to non-ssl ports and check what happens.

File: modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java [hunk 2]
```diff
@@ -78,13 +73,10 @@ private boolean readMagic(ByteBuf byteBuf) {
         if (byteBuf.readableBytes() < MAGIC_BYTES.length)
             return false;
 
-        assert data.length == MAGIC_BYTES.length;
-
-        byteBuf.readBytes(data, 0, MAGIC_BYTES.length);
+        var data = new byte[MAGIC_BYTES.length];
```
Comment (ivandasch) [inline modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java:76]: Seems that logic with `magicFailed` and `magicDecoded` works only one time. And seemd that it is not needed at all
Just add 
```
 if (byteBuff.readableBytes() < MAGIC_BYTES.length) {
            return false
 }
```

Comment (ptupitsyn) [reply modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java:76]: 1. The first 4 bytes of every client connection must contain the magic header, as per [IEP-76](https://cwiki.apache.org/confluence/display/IGNITE/IEP-76+Thin+Client+Protocol+for+Ignite+3.0)

2. New `ClientMessageDecoder` instance is created for every `Channel` (connection) thanks to `ChannelInitializer`

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/package-info.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Table API requests.
+ */
+package org.apache.ignite.client.handler.requests.table;
```
Comment (ivandasch) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/package-info.java:21]: Add new line at the end

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTupleGetRequest.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+     *
+     * @param in Unpacker.
+     * @param out Packer.
+     * @param tables Ignite tables.
+     * @return Future.
+     * @throws IOException On serialization error.
+     */
+    public static CompletableFuture<Void> process(ClientMessageUnpacker in, ClientMessagePacker out,
+                                                  IgniteTables tables)
+            throws IOException {
```
Comment (ivandasch) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTupleGetRequest.java:46]: Could you please move IOException to previous line and each input parameter at new line?
According to this rule: https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines#CodingGuidelines-MethodArguments

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTupleGetRequest.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.client.handler.requests.table;
+
+import java.io.IOException;
+import java.util.concurrent.CompletableFuture;
+
```
Comment (ivandasch) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTupleGetRequest.java:22]: Same, here and in other places, there should be no empty line between imports.

File: modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java [hunk 6]
```diff
@@ -17,48 +17,43 @@
 
 package org.apache.ignite.client.proto;
 
-import java.nio.ByteBuffer;
 import java.util.Arrays;
-import java.util.List;
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java]: Fixed

File: modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java (hunk 2 above)
Comment (ivandasch) [reply modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java:75]: Ok, sorry, I got it. I've removed original comment after realizing this

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/package-info.java (hunk 3 above)
Comment (ptupitsyn) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/package-info.java]: Fixed

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTupleGetRequest.java (hunk 5 above)
Comment (ptupitsyn) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTupleGetRequest.java]: Fixed

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTupleGetRequest.java (hunk 4 above)
Comment (ptupitsyn) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTupleGetRequest.java]: Fixed here and in some other places.

File: modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java (hunk 1 above)
Comment (ptupitsyn) [reply modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java:47]: I don't think 2 GB is too large. Ignite 2.x had the same limit and it was fine. Some users even wanted larger requests.

Comment (ivandasch) [reply modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java:47]: It should be configurable at least. For many users this is an issue

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTableCommon.java [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+    /**
+     * Reads a tuple as a map, without schema.
+     *
+     * @param unpacker Unpacker.
+     * @param table Table.
+     * @return Tuple.
+     * @throws IOException When deserialization fails.
+     */public static Tuple readTupleSchemaless(ClientMessageUnpacker unpacker, TableImpl table) throws IOException {
+        var cnt = unpacker.unpackMapHeader();
```
Comment (ivandasch) [inline modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTableCommon.java:143]: Oops something wrong with ident

Comment (ptupitsyn) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/requests/table/ClientTableCommon.java:144]: Fixed

File: modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java (hunk 1 above)
Comment (ptupitsyn) [reply modules/client-common/src/main/java/org/apache/ignite/client/proto/ClientMessageDecoder.java:48]: Added TODO with a link to https://issues.apache.org/jira/browse/IGNITE-15164

**Review by isapego — APPROVED** [review]
Looks good to me.

**Review by ivandasch — APPROVED** [review]
Looks good to me

**Pull request merged**